pkg/tidb/forwarder.go

```diff
@@ -52,7 +52,10 @@ func (f *Forwarder) Start(ctx context.Context) error {
 	f.sqlPort = f.sqlProxy.port()
 	f.statusPort = f.statusProxy.port()
 
+	// update tidb endpoints at least once when start
+	_ = f.updateTiDBEndpoints()
 	go f.pollingForTiDB()
```

Comment:
Hmm. I cannot understand the change actually.

The previous workflow is as follows:

1. go f.pollingForTiDB()
2. inside pollingForTiDB: for { ..
3. inside for: fetch a topology

The new workflow is as follows:

1. fetch a topology
2. go f.pollingForTiDB()

The only difference between the two is that, in the previous workflow the polling will be performed **after** the goroutine is started.

Reply:
I want to make sure that the topology was fetched in the fx start workflow before we ready to serve the dashboard services.

The key point is to get all required metadata or configuration stuff and then provide services. After that, we can continually update the config on demand.

Is there any better ideas? And should we do same thing to dynamic config manager?

Reply:
If we provide the service first and get the required configuration later. This doesn't seem to guarantee that the service is ready.

Reply:
> If we provide the service first and get the required configuration later. This doesn't seem to guarantee that the service is ready.

Yes. Currently they only provides "eventual consistency". TiDB Dashboard has never ensured to be ready at the moment when server is started, as user will always access TiDB Dashboard at least after a short moment (e.g. as long as several seconds).